src/frontend/src/expr/function_call.rs

```diff
@@ -175,6 +175,77 @@ impl FunctionCall {
                     datatype: Box::new(DataType::Varchar),
                 })
             }
+            ExprType::ArrayCat => {
+                if inputs.len() != 2 {
+                    return Err(ErrorCode::BindError(format!(
+                        "Function `ArrayCat` takes 2 arguments ({} given)",
+                        inputs.len()
+                    ))
+                    .into());
+                }
+                let left_type = inputs[0].return_type();
+                let right_type = inputs[1].return_type();
+
+                let wrap_single_elem = |elem, ret_type| -> ExprImpl {
+                    Self {
+                        func_type: ExprType::Array,
+                        return_type: ret_type,
+                        inputs: vec![elem],
+                    }
+                    .into()
+                };
+
+                let return_type = match (&left_type, &right_type) {
```

Comment:
As @xxchan has pointed out, `||` = `array_append` + `array_prepend` + `array_cat`. And my suggestion is:
* In `FunctionCall::new` (already refactored to `infer_type_for_special`), we just require the two operands to have the same array type. This is the `array_cat` function from PostgreSQL.
* In `bind_concat_op`, we decide whether the `||` operator is `ConcatOp` for string, or `array_cat` for two arrays of the same type, or `append/prepend` possibly backed by rewriting to `array_cat` with `wrap_single_elem`. Another example of such rewrite is from `concat` to `concat_ws`. (Do not confuse these 2 with `ConcatOp`, they are 3 different things with proc oids 3058, 3059, 1258 in `pg_proc`.)

Extra thoughts:
* `array_cat` actually support arrays of compatible types beyond same types, for example `array_cat(array[1], array[1.5])`. This can be supported via `align_types` (backed by `cast_ok_array`). But feel free to defer this support if you find bugs during integration.
* The match here have 2 arms for `append/prepend` each. It may be cleaner to wrap first and check whether they are the same (or compatible). It is just to assist resolving `||` into one of `array_cat`, `array_append` or `array_prepend` with a small overhead in binder. We are still able to avoid the overhead of wrapping in backend by implementing `array_append` and `array_prepend` directly later if desired.

To summary, the pseudo-code for `bind_concat_op` could be:
```
if align_types(left, right).is_ok() {
  array_cat(left, right)
} else if align_types(wrapped(left), right).is_ok {
  array_prepend(left, right) or array_cat(wrapped(left), right)
} else if align_types(left, wrapped(right)).is_ok {
  array_append(left, right) or array_cat(left, wrapped(right))
}
```

Reply:
Update: `array_cat` can also work between array and its element, but only when the element itself is an array and the other operand has one more dimension. Not sure why it was designed like this yet ...
https://github.com/postgres/postgres/blob/86a4dc1e6f29d1992a2afa3fac1a0b0a6e84568c/src/backend/utils/adt/array_userfuncs.c#L212-L214

Reply:
> Update: `array_cat` can also work between array and its element, but only when the element itself is an array and the other operand has one more dimension. Not sure why it was designed like this yet ... 

What's more, `array_prepend/append` only supports 1d array..

Reply:
> > Update: `array_cat` can also work between array and its element, but only when the element itself is an array and the other operand has one more dimension. Not sure why it was designed like this yet ...
> 
> What's more, `array_prepend/append` only supports 1d array..

These are checked in the latest version of this PR🤪

Reply:
In the impl of PG `array_cat`, it has branches to compute the result dimensions, but uses the same code to build the result array (just some memcpy). I guess this is feasible due to the memory layout of array. But I'm not sure whether `array_append` an array as an element would be less efficient.